### PR TITLE
Fix no carriers in checkout bad cart rules

### LINKED ISSUE #11172: No carrier available in checkout with free shipping cart rule
**EDIT**: to get the reproductible scenario, jump to https://github.com/PrestaShop/PrestaShop/issues/11172#issuecomment-502165356 ◀️ 

**Describe the bug**
When I apply a free shipping cart rule whith a code on a specific product, no carriers are available in check out, only on my production environment

**To Reproduce**
1. Update Prestashop from 1.7.4.1 to 1.7.4.2 and 1.7.4.3 with 1 click upgrade module on DEV environment.
2. Push new sources to PROD environment with git
3. Apply upgrade.php script to update database
4. Create cart rule with code and free shipping on specific country (France) and specific product
5. Order the specific product and choose a delivery address to France
6. A carrier is proposed. Apply code. Voucher is ok but no carrier are available after updating.

Steps to reproduce the issue [Khouloud]

1. Create a cart rule "Free shipping" with/without code
2. Create any cart rule without code & disable it
3. Go to the FO => add some product to the cart => complete the checkout process => in the SHIPPING METHOD step => add the code for the first cart rule (free shipping)
4. Go to BO => Enable the second cart rule
5. In step 3 => reload the page => "Unfortunately, there are no carriers available for your delivery address." is displayed

https://drive.google.com/file/d/1JWZ_iX-EpXKvwepfj9B4usDZYBBUQCBz/view
**Additionnal information**
PrestaShop version: 1.7.4.3
I haven't problem in DEV environment. If i get a dump from PROD and set in DEV, I reproduce problem.
Something must be wrong in database but I can't find what is it...

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Port of https://github.com/PrestaShop/PrestaShop/pull/12878, fix no carrier available in checkout with free shipping cart rule
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes https://github.com/PrestaShop/PrestaShop/issues/11172
| How to test?  |

-  Create cart rule without code 
- In the condition Tab "Minimum amount 100 EUR & Shipping included 
- In the actions tab, apply any action => Free shipping
- Go to FO, add to cart a product which price = 70 EUR < 100 EUR (must be < 100 EUR) => so the cart rule is not applied
- Complete your order => in step 3: SHIPPING METHOD => Unfortunately, there are no carriers available for your delivery address.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/16087)

### CODE REVIEW

**Comment by matks** [comment]
@marionf suggested we merge https://github.com/PrestaShop/PrestaShop/pull/12878 so I re-worked it by adding the behat scenario that validates it

Once we make Travis accept it, we can merge it. Although I'm worried about the possible consequences, @marionf trusts QA capabilities to spot them.

File: tests/Integration/Behaviour/Features/Scenario/Order/add_order_from_bo.feature
Comment (sarjon) [inline tests/Integration/Behaviour/Features/Scenario/Order/add_order_from_bo.feature]: :thinking:

Comment (PierreRambaud) [reply tests/Integration/Behaviour/Features/Scenario/Order/add_order_from_bo.feature]: `@ooooo`  :open_mouth:

File: tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php
Comment (matks) [inline tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php:67]: 👍

**Comment by matks** [comment]
Thanks for the help @PierreRambaud 
it seems I cannot approve the PR as I'm the author 😅

**Review by matthieu-rolland — DISMISSED** [review]
lGTM

**Comment by djbuch** [comment]
OK, this is great ! I think we are on the good way :)

**Comment by PierreRambaud** [comment]
@djbuch Eveyrthing is always on the good way when you have tests :)

**Comment by sarahdib** [comment]
@matks 
there is a problem with this PR 

If I add a product to the cart I can't delete it anymore.

If I add a cart rule and try to delete it, I need to refresh to see the change.

If I delete the cart rule and try to add it again, a warning is displayed and indicate that the cart rule is already added

**Comment by PierreRambaud** [comment]
@sarahdib Is it also happened when you don't have the PR? :thinking:

**Comment by PierreRambaud** [comment]
Errors should have been fix with the latest commit.

**Review by Progi1984 — DISMISSED** [review]
Since this concerns the file Cart.php, I ask for a second review

**Comment by Robin-Fischer-PS** [comment]
Hi @matks !

We've got a problem !

- When I add a cart rule with a gift (with code), the associated gift and cart rule are not displayed. I need to refresh to see it.

- When there is this cart rule with gift on my cart, I can't delete product unless I first delete the gift Cart Rule.

Cart rules for this test :
- "Test" : No code, applies when order amount superior to 5€, gives 2€ reduction tax included
- "Gift" : Code "GIFT", no conditions, gift "Today is a good day mug".

See screen record : 
https://drive.google.com/file/d/1eHzQtVkeiuisRtRbYCKPjCyYUgvjWGnK/view?usp=sharing

Thanks ;)

**Review by Progi1984 — APPROVED** [review]

**Comment by Robin-Fischer-PS** [comment]
Hi @colinegin and Product Team !

There is have a small change of behavior with last fix on this PR, when I have a gift cart rule in a cart.

- Without this PR (1.7.7) : 

I cannot delete a gifted product from the cart, I need to delete the cart rule. This may "surprise" the customer, who might wonder why he can't delete the product.

- With this PR : 

I can either delete the gifted product from the cart or directly delete the cart rule. But if I delete from the cart, the cart rule is still displayed, with a "0,00€" discount.
The cart rule can still be deleted, and added afterwards.

pr16087 gift deletion

Is this OK ?

Thanks ;)

**Comment by colinegin** [comment]
Hello @Robin-Fischer-PS ,

After reviewing with you the new behaviour, I confirm this change is needed to improve user experience and get free gifts removable in the cart. 

All good for me !

Thanks ;)

**Comment by matks** [comment]
Thank you @PierreRambaud @djbuch @matthieu-rolland @Robin-Fischer-PS @Progi1984

**Comment by PierreRambaud** [comment]
Thanks @matks :)

**Pull request merged**
